### PR TITLE
76749 ErpFacade endpoint to receive callback from SAP (Part 1)

### PR DESCRIPTION
**Pull request by Nadeem14456**
1. Do follwing things whn ERP Facade receives callback from SAP -
   - Validate and check if blob exists in storage account for given traceId/CorrelationId.
   - Update ResponseDateTime in azure table against same traceId/CorrelationId.
2.  Added Unit Test cases

Please note that this endpoint is partially implemented as part of this PBI. Hence, automation test cases will be added when we complete this endpoint with price information mapping in next sprint.

### CODE REVIEW

File: src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs [hunk 1]
```diff
@@ -7,20 +10,48 @@ namespace UKHO.ERPFacade.API.Controllers
     public class ErpFacadeController : BaseController<ErpFacadeController>
     {
         private readonly ILogger<ErpFacadeController> _logger;
+        private readonly IAzureTableReaderWriter _azureTableReaderWriter;
+        private readonly IAzureBlobEventWriter _azureBlobEventWriter;
+
+        public const string TRACEIDKEY = "data.traceId";
```
Comment (jrippington) [inline src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs:16]: Should be TraceIdKey

File: src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs [hunk 2]
```diff
@@ -7,20 +10,48 @@ namespace UKHO.ERPFacade.API.Controllers
...
+        private const string UPDATE_RESPONSE_TIME = "ResponseDateTime";
```
Comment (jrippington) [inline src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs:17]: Should be UpdateResponseTime

File: src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs [hunk 3]
```diff
@@ -7,20 +10,48 @@ namespace UKHO.ERPFacade.API.Controllers
...
 
         [HttpPost]
-        [Route("/erpfacade/getpriceinfo")]
-        public virtual async Task<IActionResult> Post()
+        [Route("/erpfacade/priceinformation")]
+        public virtual async Task<IActionResult> Post([FromBody] JObject requestJson)
         {
-            await Task.CompletedTask;
-            return Ok();
+            string traceId = requestJson.SelectToken(TRACEIDKEY)?.Value<string>();
```
Comment (jrippington) [inline src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs:35]: Use var

File: src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs [hunk 4]
```diff
@@ -7,20 +10,48 @@ namespace UKHO.ERPFacade.API.Controllers
...
+
+            if (string.IsNullOrEmpty(traceId))
+            {
+                _logger.LogWarning(EventIds.TraceIdMissingInSAPEvent.ToEventId(), "TraceId is missing in the event received from the SAP.");
+                return new BadRequestObjectResult(StatusCodes.Status400BadRequest);
+            }
+
+            await _azureTableReaderWriter.UpdateEntity(traceId, UPDATE_RESPONSE_TIME);
+
+            bool isBlobExists = _azureBlobEventWriter.CheckIfContainerExists(traceId);
```
Comment (jrippington) [inline src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs:45]: Use var

File: src/UKHO.ERPFacade.Common/IO/AzureBlobEventWriter.cs [hunk 5]
```diff
@@ -44,5 +48,15 @@ private BlobClient GetBlobClient(string containerName)
             BlobClient blobClient = blobContainerClient.GetBlobClient(blobName);
             return blobClient;
         }
+
+        public bool CheckIfContainerExists(string containerName)
+        {
+            BlobServiceClient serviceClient = new(_azureStorageConfig.Value.ConnectionString);
+
+            var container = serviceClient.GetBlobContainerClient(containerName);
+            bool isExists = container.Exists();
```
Comment (jrippington) [inline src/UKHO.ERPFacade.Common/IO/AzureBlobEventWriter.cs:57]: Use var

File: src/UKHO.ERPFacade.Common/IO/AzureBlobEventWriter.cs [hunk 6]
```diff
@@ -14,6 +17,7 @@ public class AzureBlobEventWriter : IAzureBlobEventWriter
     {
         private readonly ILogger<AzureBlobEventWriter> _logger;
         private readonly IOptions<AzureStorageConfiguration> _azureStorageConfig;
+        private Task<bool> isExists;
```
Comment (jrippington) [inline src/UKHO.ERPFacade.Common/IO/AzureBlobEventWriter.cs:20]: Not sure what this is for? Please remove if not used - if it is, please rename to indicate what it does, and use _ prefix convention

File: src/UKHO.ERPFacade.Common/IO/AzureTableReaderWriter.cs [hunk 7]
```diff
@@ -16,6 +16,8 @@ public class AzureTableReaderWriter : IAzureTableReaderWriter
     {
         private readonly ILogger<AzureTableReaderWriter> _logger;
         private const string ERP_FACADE_TABLE_NAME = "eesevents";
+        private const string UPDATE_REQUEST_TIME = "RequestDateTime";
```
Comment (jrippington) [inline src/UKHO.ERPFacade.Common/IO/AzureTableReaderWriter.cs:19]: Should be UpdateRequestTime

File: src/UKHO.ERPFacade.Common/IO/AzureTableReaderWriter.cs [hunk 8]
```diff
@@ -16,6 +16,8 @@ public class AzureTableReaderWriter : IAzureTableReaderWriter
...
+        private const string UPDATE_RESPONSE_TIME = "ResponseDateTime";
```
Comment (jrippington) [inline src/UKHO.ERPFacade.Common/IO/AzureTableReaderWriter.cs:20]: Should be UpdateResponseTime

File: src/UKHO.ERPFacade.Common/IO/AzureTableReaderWriter.cs [hunk 9]
```diff
@@ -16,6 +16,8 @@ public class AzureTableReaderWriter : IAzureTableReaderWriter
     {
         private readonly ILogger<AzureTableReaderWriter> _logger;
         private const string ERP_FACADE_TABLE_NAME = "eesevents";
```
Comment (jrippington) [inline src/UKHO.ERPFacade.Common/IO/AzureTableReaderWriter.cs:18]: Should be ErpFacadeTableName

File: src/UKHO.ERPFacade.Common/IO/AzureTableReaderWriter.cs [hunk 10]
```diff
@@ -74,6 +76,24 @@ await foreach (var entity in entities)
             return records.FirstOrDefault();
         }
 
+        public async Task UpdateEntity(string traceId, string updateColumn)
+        {
+            TableClient tableClient = GetTableClient(ERP_FACADE_TABLE_NAME);
```
Comment (jrippington) [inline src/UKHO.ERPFacade.Common/IO/AzureTableReaderWriter.cs:81]: Use var

File: src/UKHO.ERPFacade.Common/IO/AzureTableReaderWriter.cs [hunk 11]
```diff
@@ -74,6 +76,24 @@ await foreach (var entity in entities)
             return records.FirstOrDefault();
         }
 
+        public async Task UpdateEntity(string traceId, string updateColumn)
```
Comment (jrippington) [inline src/UKHO.ERPFacade.Common/IO/AzureTableReaderWriter.cs:79]: Need unit tests for this method

**Review by jrippington — CHANGES_REQUESTED** [review]

File: src/UKHO.ERPFacade.Common/Logging/EventIds.cs [hunk 12]
```diff
@@ -54,5 +54,21 @@ public enum EventIds
         /// 940013 - Data pushed to SAP
         /// </summary>
         DataPushedToSap = 940013,
+        /// <summary>
+        /// 940017 - Updated entity successfully in Azure Table
```
Comment (harshal11869) [inline src/UKHO.ERPFacade.Common/Logging/EventIds.cs:58]: Please correct he sequence number.

File: src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs [hunk 13]
```diff
@@ -7,20 +10,48 @@ namespace UKHO.ERPFacade.API.Controllers
...
         {
             _logger = logger;
+            _azureTableReaderWriter = azureTableReaderWriter;
+            _azureBlobEventWriter = azureBlobEventWriter;
         }
 
         [HttpPost]
-        [Route("/erpfacade/getpriceinfo")]
-        public virtual async Task<IActionResult> Post()
+        [Route("/erpfacade/priceinformation")]
```
Comment (harshal11869) [inline src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs:32]: Please add [Authorize] attribute on call and [Authorize(Policy = "PriceInformationApiCaller")] on endpoint.

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs (hunk 2 above)
Comment (Nadeem14456) [reply src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs:17]: updated.

File: src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs (hunk 1 above)
Comment (Nadeem14456) [reply src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs:16]: updated.

File: src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs (hunk 3 above)
Comment (Nadeem14456) [reply src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs:35]: done.

File: src/UKHO.ERPFacade.Common/IO/AzureBlobEventWriter.cs (hunk 6 above)
Comment (Nadeem14456) [reply src/UKHO.ERPFacade.Common/IO/AzureBlobEventWriter.cs:20]: done.

File: src/UKHO.ERPFacade.Common/IO/AzureTableReaderWriter.cs (hunk 8 above)
Comment (Nadeem14456) [reply src/UKHO.ERPFacade.Common/IO/AzureTableReaderWriter.cs:20]: done.

File: src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs (hunk 13 above)
Comment (Nadeem14456) [reply src/UKHO.ERPFacade.API/Controllers/ErpFacadeController.cs:32]: this change will be done in upcoming sprints.

File: src/UKHO.ERPFacade.Common/IO/AzureTableReaderWriter.cs (hunk 11 above)
Comment (Nadeem14456) [reply src/UKHO.ERPFacade.Common/IO/AzureTableReaderWriter.cs]: Done. This method is separated for UpdateRequestTime & UpdateResponseTime.

**Review by martyn-fewtrell — APPROVED** [review]
Looks good

**Pull request merged**